### PR TITLE
William/scroll performance

### PR DESCRIPTION
**Pull request by whoiswillma**
- The search bar and filter bar are in their own section, instead of being added directly to the `collectionView`.
  * This was done to make the search and filter bar stop "jiggling" when scrolling the `collectionView`.
- The AppDev logo reached by pulling down on the `collectionView` is now instantiated in the `EateriesSharedViewController` to stop each `EateryViewController` from needing to add their own copy of it.

### CODE REVIEW

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (young-k) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:208]: I don't think we should have `\n` after a comment, I don't think it adds any clarity!

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (young-k) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:361]: Can't this `for` loop and the above `for view in views` be combined?

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (young-k) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:468]: might be cleaner with a ternary, but this is personal preference –– will leave up to you!

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (young-k) [inline Eatery/Controllers/Eateries/EateriesViewController.swift]: Can't this comment fit in two lines // is there a reason for splitting it up?

Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift]: I use 80 columns as a soft guide, but I'm up for bumping that to 120

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:361]: Certainly, but I think two for-loops reads better because they distinguish the setup from the animations.

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (young-k) [reply Eatery/Controllers/Eateries/EateriesViewController.swift]: cc @kevinchan159

**Comment by whoiswillma** [comment]
I noticed an issue where entering search text would cause the keyboard to dismiss. Because `reloadData` was being called in the `reloadEateries` method, the search bar would be removed from the window and then reattached, causing it to lose its first responder status. 

Instead of trying to track which sections to insert and remove, I made the number of sections constant and just hide and show them as needed, which is simpler and also stops the search and filter bar cell from being reloaded

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:349]: nit: pref to use `forEach`

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:355]: same here

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:361]: i agree w/ young here. you could just add comments inside the for loop but if you feel strongly about this you can keep it

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:468]: pref ternary here too

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/Eateries/EateriesViewController.swift]: nit: could remove the `else` since you're returning on line 590

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/Eateries/EateriesViewController.swift]: same here

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/Eateries/EateriesViewController.swift]: same here

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
looks great! just left some small nits

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/Eateries/EateriesViewController.swift]: Lets go w/ 100 characters as the char limit per line. Thats the recommended limit in the Google/Airbnb style guides

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:349]: Why in general is it preferable to use `forEach` over the regular `for` statements?

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:361]: I don't feel particularly strong about this; It's more a case of it was that way before (which is not a good reason in-and-of-itself, but which I think is reasonable if I came to the code like it is) in combination with (what I perceive as) the additional readability.

It's more so that the `for`-loops behave as separate paragraphs in an essay: they serve different purposes, therefore they should be split, even if semantically having one and having two are the same.

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:349]: There isnt really a benefit of using one over the other besides consistency. Since Swift is functional and we try to use functional functions like `map/filter/reduce` where possible, I think using `forEach` where possible makes sense

**Review by kevinchan159 — APPROVED** [review]
approving to unblock but address comments before merging pls!

**Comment by kevinchan159** [comment]
small note: Could you click on `Resolve conversation` after you finish addressing a comment? Just makes it a bit easier for reviewers to know what you've addressed

File: Eatery/Controllers/Eateries/EateriesCollectionViewGridLayout.swift
Comment (chalo2000) [inline Eatery/Controllers/Eateries/EateriesCollectionViewGridLayout.swift:25]: Perhaps consider replacing this with a ternary

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:179]: Remove print statement

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/Eateries/EateriesViewController.swift:462]: Why are there 4 sections if our switch above only has cases for 3 sections and we return a 1 whenever section == 0  below? Something seems off to me with this.

**Review by chalo2000 — COMMENTED** [review]
Just tiny fixes and a question about the sections in the collectionView

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:349]: I see three main reasons for sticking with "regular" `for` loops with Eatery:
1. Eatery already uses `for` almost everywhere, so being consistent is good (when it makes sense)
2. `forEach` produces `Void` as its value, whereas `map` (and the other functional methods in Swift) produce meaningful values. This matters because `for` loops are control flow statements, and should therefore be distinguished from expressions that produce values. Instead of having to read the words `forEach`, `for` loops are separated visually from the rest of the code, by the syntax itself as well as by syntax highlighting. 
3. `for` loops are generally used to manipulate state, whereas `map` is expected to _not_  manipulate state. Similarly to 2., as human beings who have to look at code with our eye-balls, having this separation is useful for distinguishing the purpose of code.

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:179]: Hmmn, I feel keeping this one is worthwhile. It doesn't get printed that often and it's also a pretty significant event.

File: Eatery/Controllers/Eateries/EateriesViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesViewController.swift:462]: The search and filter bars are now a cell in the collection view. In order to make Some Things™ easier, I put them in their own section. Now there are four sections: the "bars" section which only ever contains one cell, and then the remaining three sections ("favorites", "open", and "closed") as usual.

File: Eatery/Controllers/Eateries/EateriesCollectionViewGridLayout.swift
Comment (whoiswillma) [reply Eatery/Controllers/Eateries/EateriesCollectionViewGridLayout.swift:25]: I think a ternary would be a bit too unwieldy: the conditional and both expressions are pretty long.

**Review by chalo2000 — APPROVED** [review]

**Pull request merged**
